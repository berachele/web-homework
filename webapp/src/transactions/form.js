import React, { useState } from 'react'

function Form(props){
    const [formValues, setFormValues] = useState(props.formValues)

    const handleChange = event => {
        setFormValues({
          ...formValues,
          [event.target.name]: event.target.value
        })
      }

    const onSubmit = event => {
        //stops the rerendering after adding transaction
        event.preventDefault()

        props.onSubmit(formValues.id, formValues.user, formValues.merchant, formValues.amount, formValues.date, formValues.isOpen)
        //clearing form after submission
        setFormValues(props.initialFormValues)
    }

    return (
      <form onSubmit={onSubmit}>
          <label> User: &nbsp;
            <input
              data-testid='user'
              onChange={handleChange} 
              value={formValues.user}
              name='user'
              type='text'
            />
          </label>
          <br/>

          <label> Merchant: &nbsp;
            <input
              data-testid='merchant'
              onChange={handleChange} 
              value={formValues.merchant}
              name='merchant'
              type='text'
            />
          </label>
          <br/>

          <label> Amount: &nbsp;$
            <input
              data-testid='amount'
              className='amount'
              onChange={handleChange} 
              value={formValues.amount}
              name='amount'
              type='text'
            />
          </label>
          <br/>
          
          <label> Date: &nbsp;
            <input
              data-testid='date'
              onChange={handleChange} 
              value={formValues.date}
              name='date'
              type='text'
              placeholder='MM/DD/YYYY'
            />
          </label>
          <br /><br />
  
          <input type='submit' data-testid='submit' className='submit'/>
        </form>
    )
  }

  export default Form
  