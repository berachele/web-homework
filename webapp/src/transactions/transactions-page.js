import React, { useState } from 'react'
import { v4 as uuid } from 'uuid'
//components
import Form from './form'
//materialui icons
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';

//dummy data to show transactions on Mount
const initialTransactions = [
  { id: uuid(), user: 'Hermoine', merchant: 'Honey Duke\'s', amount: '$100', date: '12/20/2020', isOpen: false },
  { id: uuid(), user: 'Harry', merchant: 'Ollivander\'s Wand Shop', amount: '$5', date: '01/05/2021', isOpen: false },
]

const initialFormValues = {
  //id with uuid() only being used when you create a transaction
  id: uuid(),
  user: '',
  merchant: '',
  amount: '',
  date: '',
  isOpen: false
}

export function Transactions () {
  const [transactions, setTransactions] = useState(initialTransactions)
  const [formValues, setFormValues] = useState(initialFormValues)


  const createTransactionSubmission =  (id, user, merchant, amount, date ) => {
    const newTransaction = { id, user, merchant, amount, date }
    setTransactions([...transactions, newTransaction])
  }

  const editTransactionSubmission = ( id, user, merchant, amount, date ) => {
    //setting isOpen to false so form will close after submission
    const newTransaction = { id, user, merchant, amount, date, isOpen: false}

    let newTransactionIndex = transactions.findIndex((transaction) => {
      return transaction.id === id
    })

    let newArray = [...transactions]
    newArray[newTransactionIndex] = newTransaction
    setTransactions(newArray)
  }

  const deleteTransaction = (transaction) => {
    let transactionIndex = transactions.indexOf(transaction)
    transactions.splice(transactionIndex, 1)
    setTransactions([...transactions])
  }
  
  const openEditFrom = (transaction) => {
    let transactionIndex = transactions.indexOf(transaction)
    let newArray = [...transactions]
    newArray[transactionIndex].isOpen = true
    setTransactions(newArray)
  }

  return (
    <div>
      <Form
        onSubmit={createTransactionSubmission}
        formValues={formValues}
        initialFormValues={initialFormValues}
      />

      <h3>List of Transactions</h3>
      {
        transactions.map(transaction => 
        <>
          <div key={transaction.id} style={{ display: 'flex', alignItems: 'center' }}> 
            <p>{transaction.user} spent {transaction.amount} at {transaction.merchant} on {transaction.date}</p>
            <EditIcon onClick={()=>openEditFrom(transaction)}/> 
            <DeleteIcon onClick={()=>deleteTransaction(transaction)} style={{color: 'red'}}/> 
          </div> 
          {/* edit form will only open when transaction.isOpen is true */}
          { transaction.isOpen && 
          <Form
            onSubmit={editTransactionSubmission}
            formValues={{ 
              id: transaction.id, 
              amount: transaction.amount, 
              date: transaction.date, 
              user: transaction.user, 
              merchant: transaction.merchant, 
              isOpen: transaction.isOpen 
            }}
          />
          }
        </>)
      }
    </div>
  )
}
