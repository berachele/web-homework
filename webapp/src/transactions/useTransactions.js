//custom hook for passing state for transactions to be passable to testing file and AppRouter
import { useState } from 'react'


//dummy data to pass into props (useState's transactions and setTransactions)
let initialMockTransactions = [
    {
      id: "4ea33565-a98f-4151-846a-0a35d4da9a03",
      user: "Hermoine",
      merchant: "Flourish and Blotts",
      amount: "100",
      date: "12/20/2020",
      isOpen: false,
    },
    {
      id: "8cbfc4d1-b506-4755-b18e-1765ff29fe71",
      user: "Harry",
      merchant: "Ollivander's Wand Shop",
      amount: "5",
      date: "01/05/2021",
      isOpen: false,
    }
  ]
  
function useTransactions() {
  const [mockTransactions, setMockTransactions] = useState(initialMockTransactions)

  return { mockTransactions, setMockTransactions }
}

export default useTransactions