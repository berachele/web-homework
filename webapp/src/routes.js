import React from 'react'
import { BrowserRouter as Router, Route, Link } from 'react-router-dom'
import { css } from '@emotion/core'
import { Home } from './home'
import { Transactions } from './transactions'

function AppRouter () {
  return (
    <Router>
      <div css={layoutStyle}>
        <nav css={navStyle}>
          <ul >
            <li>
              <Link to='/'>Home</Link>
            </li>
            <li>
              <Link to='/transactions'>Transactions</Link>
            </li>
          </ul>
        </nav>
        <div className='main-content' css={contentStyle}>
          <Route component={Home} exact path='/' />
          <Route component={Transactions} exact path='/transactions' />
        </div>
      </div>
    </Router>
  )
}

export default AppRouter

const layoutStyle = css`
    box-sizing: border box;
    margin:0;
    padding:0;
    max-width:100%;
`

const navStyle = css`
  grid-row: 1;
  background-color: grey;
  padding: 2%;

  & > ul {
      display: flex;
      flex-direction: row;
      list-style-type: none;
  }
  
  & > ul > li:not(:first-child) {
    margin-left: 16px;
  }
`

const contentStyle = css`
  grid-row: 2;
`
