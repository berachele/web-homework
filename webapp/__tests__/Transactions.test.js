import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Transactions } from '../src/transactions';

//dummy data to pass into props (useState's transactions and setTransactions)
let mockTransactions = [
  {
    id: "4ea33565-a98f-4151-846a-0a35d4da9a03",
    user: "Hermoine",
    merchant: "Flourish and Blotts",
    amount: "100",
    date: "12/20/2020",
    isOpen: false,
  },
  {
    id: "8cbfc4d1-b506-4755-b18e-1765ff29fe71",
    user: "Harry",
    merchant: "Ollivander's Wand Shop",
    amount: "5",
    date: "01/05/2021",
    isOpen: false,
  }
]

describe('Transactions page test suite', () => {
  it('renders "List of Transactions" message', () => {
    const { getByText } = render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)
    const listTitle = getByText('List of Transactions')
    
    expect(listTitle).toBeInTheDocument()
  })

  // this will also watch Form because Form is imported in Transactions
  // it('matches Transaction component snapshot', () => {
  //   const tree = renderer.create(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>).toJSON()

  //   expect(tree).toMatchSnapshot()
  // })

  it('Testing if inputs on form are visible on add form', () => {
    const { getByLabelText } = render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)

    const userInput = getByLabelText(/User/i)
    const merchantInput = getByLabelText(/Merchant/i)
    const amountInput = getByLabelText(/Amount/i)
    const dateInput = getByLabelText(/Date/i)

    expect(userInput.value).toBe('')
    expect(merchantInput.value).toBe('')
    expect(amountInput.value).toBe('')
    expect(dateInput.value).toBe('')
  })

  it('Form to enter and add transaction successfully works', async() => {
    const { getByTestId } = render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)
    //entering values
    const userInput = getByTestId(/user/i)
    const merchantInput = getByTestId(/merchant/i)
    const amountInput = getByTestId(/amount/i)
    const dateInput = getByTestId(/date/i)

    fireEvent.change(userInput, { target: { value: 'Ron' } })
    fireEvent.change(merchantInput, { target: { value: 'Weasley\'s Wizard Wheezes' } })
    fireEvent.change(amountInput, { target: { value: '15' } })
    fireEvent.change(dateInput, { target: { value: '07/31/2020' } })
    
    expect(userInput.value).toBe('Ron')
    expect(merchantInput.value).toBe('Weasley\'s Wizard Wheezes')
    expect(amountInput.value).toBe('15')
    expect(dateInput.value).toBe('07/31/2020')
    //submission
    fireEvent.click(getByTestId(/submit/i))

    const userValue = await userInput.value
    const merchantValue = await merchantInput.value
    const amountValue = await amountInput.value
    const dateValue = await dateInput.value

    expect(userValue).toBe('')
    expect(merchantValue).toBe('')
    expect(amountValue).toBe('')
    expect(dateValue).toBe('')
  })

  it('Clicking Edit Icon successfully opens up edit form', async() => {
    render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)
    const editBtnNodeList = document.querySelectorAll('#edit')
    const firstTransaction = editBtnNodeList[0]
    const formsList = document.querySelectorAll('.form')
    const formListLength = formsList.length

    expect(formListLength).toBe(1)

    fireEvent.click(firstTransaction)

    const newFormsList = await document.querySelectorAll('.form')

    expect(newFormsList.length).toBe(2)
  })

  it('Form to edit transaction successfully works', async() => {
    const { getByDisplayValue } = render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)
    //editing values
    const amountInput = getByDisplayValue('100')

    fireEvent.change(amountInput, { target: { value: '45' } })
    
    const amountValue = await amountInput.value

    expect(amountValue).toBe('45')
    
    //submission
    const submitButtons = document.querySelectorAll('.submit')
    const mySubmitButton = submitButtons[1]

    fireEvent.click(mySubmitButton)

    const updatedTransaction = await document.querySelector('.transaction').textContent

    expect(updatedTransaction).toBe('Hermoine spent $45 at Flourish and Blotts on 12/20/2020')
  })
})

  it('Clicking Delete Icon successfully deletes that transaction', async() => {
    render(<Transactions transactions={mockTransactions} setTransactions={() => {}}/>)
    const deleteBtnNodeList = document.querySelectorAll('#delete')
    const nodeListLength = deleteBtnNodeList.length
    const firstTransaction = deleteBtnNodeList[0]

    expect(nodeListLength).toBe(2)

    fireEvent.click(firstTransaction)

    const newNodeListLength = await document.querySelectorAll('.transaction').length
    
    expect(newNodeListLength).toBe(1)
  })
