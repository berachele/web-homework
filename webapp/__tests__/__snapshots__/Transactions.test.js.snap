// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions page test suite matches Transaction component snapshot 1`] = `
.emotion-0 {
  border: 2px solid #9a8c98;
  background: rgba(154,140,152,.6);
  border-radius: 1%;
  width: 50%;
  padding: 1%;
}

.emotion-0>h2 {
  margin: 0;
}

.emotion-0:nth-child(2) {
  width: 27%;
}

.emotion-0:nth-child(3) {
  margin-top: 4%;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-top: 1%;
}

.emotion-1>label {
  font-weight: bold;
  padding: 1% 0;
  font-size: larger;
}

.emotion-1>label>input {
  height: 20px;
  font-size: initial;
}

.emotion-1>.submit {
  height: 24px;
  font-size: medium;
  outline: none;
  background: #169f6e;
  color: #f2e9e4;
  border: 1px solid #117E58;
  box-shadow: 1px 1px grey;
  margin-left: 14%;
}

.emotion-1>.submit:hover {
  background: #18B47D;
}

.emotion-3 {
  padding-bottom: 2%;
}

.emotion-4 {
  border-bottom: 1px solid #4A4E69;
  background: rgba(242,233,228,.5);
  border-radius: 1%;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6:hover {
  color: #474747;
}

.emotion-7 {
  color: #C4233D;
}

.emotion-7:hover {
  color: #DA2F4B;
}

<div>
  <h1>
    Transactions
  </h1>
  <div
    className="emotion-0"
  >
    <h2>
      Enter a Transaction
    </h2>
    <form
      className="form emotion-1"
      onSubmit={[Function]}
    >
      <label>
         User:  
        <input
          data-testid="user"
          name="user"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <br />
        
      <label>
         Merchant:  
        <input
          data-testid="merchant"
          name="merchant"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <br />
      <label>
         Amount:  $
        <input
          className="amount"
          data-testid="amount"
          name="amount"
          onChange={[Function]}
          type="text"
          value=""
        />
      </label>
      <br />
        
      <label>
         Date:  
        <input
          data-testid="date"
          name="date"
          onChange={[Function]}
          placeholder="MM/DD/YYYY"
          type="text"
          value=""
        />
      </label>
      <br />
      <br />
      <input
        className="submit"
        data-testid="submit"
        type="submit"
      />
    </form>
  </div>
  <div
    className="emotion-0"
  >
    <h2
      className="emotion-3"
    >
      List of Transactions
    </h2>
    <div
      className="emotion-4"
      id="transaction-block"
    >
      <div
        className="emotion-5"
      >
        <p
          className="transaction"
        >
          Hermoine
           spent $
          100
           at 
          Flourish and Blotts
           on 
          12/20/2020
        </p>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root emotion-6"
          focusable="false"
          id="edit"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root emotion-7"
          focusable="false"
          id="delete"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
    </div>
    <div
      className="emotion-4"
      id="transaction-block"
    >
      <div
        className="emotion-5"
      >
        <p
          className="transaction"
        >
          Harry
           spent $
          5
           at 
          Ollivander's Wand Shop
           on 
          01/05/2021
        </p>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root emotion-6"
          focusable="false"
          id="edit"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
          />
        </svg>
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root emotion-7"
          focusable="false"
          id="delete"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
